[MRG+1] fix StratifiedShuffleSplit train and test overlap

lesteve opened the pull request:
Fix #6121.

lesteve commented on sklearn/cross_validation.py:
```diff
@@ -1010,14 +1010,19 @@ def _iter_indices(self):
                                                 minlength=len(self.y)) == 0,
                                        )[0]
                 missing_idx = rng.permutation(missing_idx)
-                train.extend(missing_idx[:(self.n_train - len(train))])
```
The crux of the problem is that `self.n_train - len(train)` can be negative if you are unlucky with the rounding, which causes a problem in the slicing below.

lesteve commented on sklearn/cross_validation.py:
```diff
@@ -1010,14 +1010,19 @@ def _iter_indices(self):
             # dividers of the number of elements per class), we may end
             # up here with less samples in train and test than asked for.
-            if len(train) < self.n_train or len(test) < self.n_test:
+            if len(train) + len(test) < self.n_train + self.n_test:
```
This condition is more correct since you can have `len(test) < self.n_test` but no all indices have already been assigned to either train or test.

lesteve commented on sklearn/cross_validation.py:
```diff
@@ -1010,14 +1010,19 @@ def _iter_indices(self):
+                n_missing_train = self.n_train - len(train)
+                n_missing_test = self.n_test - len(test)
+
+                if n_missing_train > 0:
```
This is a slightly ad-hoc fix. Maybe it would be better to improve the logic [there](https://github.com/lesteve/scikit-learn/blob/fix-stratified-shuffle-split-train-test-overlap/sklearn/cross_validation.py#L996)

TomDLT commented:
LGTM

rvraghav93 commented:
I think this be included in the `model_selection` module too... Thanks for the fix...

TomDLT commented:
good point. [sklearn/model_selection/_split.py](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/model_selection/_split.py#L1113)

lesteve commented:
> good point. sklearn/model_selection/_split.py

I wasn't aware of that. Just curious, are model_selection.StratifiedShuffleSplit and cross_validation.StratifiedShuffle split supposed to be just a copy and paste?

rvraghav93 commented:
For the most part yes!

GaelVaroquaux commented:
> model_selection.StratifiedShuffleSplit and cross_validation.StratifiedShuffleSplit

Everything in cross_validation is deprecated. As we changed API we had to
create a new import path, in order not to break everything for people.

lesteve commented:
OK I just pushed a commit with the fix for both model_selection.StratifiedShuffleSplit and cross_validation.StratifiedShuffleSplit.

>> Just curious, are model_selection.StratifiedShuffleSplit and cross_validation.StratifiedShuffle split supposed to be just a copy and paste of each other?

> For the most part yes!

This is slightly surprising as manually maintaining consistency between the two for two more major releases may be a bit burdensome. I guess this was discussed in details though and it was deemed a better solution than, say, sharing the common code between the two.

rvraghav93 commented:
> may be a bit burdensome.

Indeed ;( An attempt to reuse the code from model selection has been made at #5568 Reviews would be welcome :grin:

rvraghav93 commented:
And thanks for extending the fix to `model_selection`

ogrisel commented on sklearn/model_selection/_split.py:
```diff
@@ -1050,7 +1050,7 @@ class StratifiedShuffleSplit(BaseShuffleSplit):
     >>> sss = StratifiedShuffleSplit(n_iter=3, test_size=0.5, random_state=0)
     >>> sss.get_n_splits(X, y)
-    3
+    3n
```
typo?

ogrisel replied:
I wonder why the docstring tests pass...

lesteve replied:
I fixed the typo and I'll take a closer look at why the docstring didn't seem to be run.

lesteve replied:
> why the docstring didn't seem to be run.

Seems like nosetests ignores files starting with an underscore. Maybe we should use `ignore-files=DONTIGNOREANYFILES` as was mentioned [here](http://stackoverflow.com/questions/17863992/force-nosetests-to-find-doctests-in-modules-starting-with-underscore)

lesteve replied:
For some reason you have to use `ignore-files=setup.py` because trying to import sklearn/cluster/setup.py confuses the hell out of nose not sure why.

ogrisel commented:
Besides the typo in the doctest and the fact that travis is still happy about it, +1 for merge as well.

amueller commented:
> This is slightly surprising as manually maintaining consistency between the two for two more major releases may be a bit burdensome

Well, there will be no changes to the deprecated code, and hopefully there are not too many bugs to fix ;)

amueller commented:
whatsnew entry?

amueller commented:
I don't understand line 997. Why ``np.minimum(cls_count - n_i, np.round(self.n_test * p_i).astype(int))``?
Shouldn't ``n_i + t_i = cls_count``?
(also, comments to explain the single-letter n_i and t_i would be kinda nice).

amueller commented:
the logic is kinda tricky now. Can you add comments about the one-letter variables please?
Any ideas if this can be made more simple?

lesteve commented:
> whatsnew entry?

I'll add a whatsnew entry. Just for the record this overlap between train and test happens only in some unlikely edge cases. You have to be pretty unlucky to hit it. Basically you have to
1. fall in the unassigned sample logic (`n_train * p_i` is of the form `2n + 0.5` with n being an integer)
2. assign too many samples to train (`n_train * p_i` is of the form `2n + 1.5`)

You also need 1. and 2. to coordinate a in a specific fashion as well.

> the logic is kinda tricky now

don't think the PR makes the logic trickier than it already was.

> Can you add comments about the one-letter variables please?
Any ideas if this can be made more simple? (otherwise lgtm)

I was thinking about whether there was a way to simplify the code, in particular by getting rid of the  unassigned samples logic,  but I wasn't able to find one right away. I can try to look at it in more details and open a separate PR if I managed to get anywhere.

lesteve committed:
FIX StratifiedShuffleSplit train and test overlap

in some edge cases and added test.

Fix was applied in both sklearn.model_selection and
sklearn.cross_validation.

lesteve commented:
> I'll add a whatsnew entry

whatsnew entry added.

ogrisel commented:
> Can you add comments about the one-letter variables please?

Which variables?

lesteve commented:
> Which variables?

Not sure but I am guessing @amueller meant `p_i`, `n_i` and `t_i` [here](https://github.com/lesteve/scikit-learn/blob/fix-stratified-shuffle-split-train-test-overlap/sklearn/cross_validation.py#L994). Since I am not changing this code, I propose to do that in a separate PR, especially if I can find a way to simplify the logic as I mentioned above.

amueller commented:
I did mean ``p_i``, ``n_i`` and ``t_i``. And yeah, the logic didn't become more complex. I was just hoping against hope it could be less complex ;)

amueller commented:
Maybe we can make that a programming challenge lol

lesteve commented:
> I did mean p_i, n_i and t_i. And yeah, the logic didn't become more complex. I was just hoping against hope it could be less complex ;)

I'll try to take a closer look at this in the next few days.

amueller commented:
I think we could try to fix ``n_i`` and ``t_i`` before the are used so that ``n_train - n_i.sum() == 0``.
We "just" need to put some samples from ``class_counts - n_i - t_i`` into ``n_i`` and ``t_i``.
I haven't found an idiomatic way to sample part of ``class_counts - n_i - t_i``, though.

``np.random.permutation(np.repeat(range(n_classes), class_counts - n_i - t_i))`` is the best I can do.
